lager/detail/nodes.hpp

```diff
@@ -200,17 +202,24 @@ class reader_node : public reader_node_base
     void collect()
     {
         using namespace std;
-        children_.erase(remove_if(begin(children_),
-                                  end(children_),
-                                  mem_fn(&weak_ptr<reader_node_base>::expired)),
-                        end(children_));
+        if (garbage_) {
+            children_.erase(
+                remove_if(begin(children_),
+                          end(children_),
+                          mem_fn(&weak_ptr<reader_node_base>::expired)),
+                end(children_));
+            garbage_ = false;
+        }
     }
 
-    int flags_ = no_flag;
     value_type current_;
     value_type last_;
     std::vector<std::weak_ptr<reader_node_base>> children_;
     signal_type observers_;
+
+    char flags_     = no_flag;
+    bool outermost_ = true;
+    bool garbage_   = false;
```

Comment:
Not fan of having 3 different flag types, i'd like to keep state as small as possible. I believe that:

- `garbage_` we can just get rid of. In the worst case we would just collect garbage one notification later. It's no problem to deal with it the way we did, IMHO.
- `outermost_` can be just one more state in `flags_`.

Reply:
so you're saying go back to just the local variable for `garbage` and only execute it for the outermost call?

Reply:
so after reviewing all the options for flags, it feels like we don't get much benefit from packing them into a single integer, and having them as separate variables increases readability.